keystone/gauss_fit.py

```diff
@@ -0,0 +1,230 @@
+
+	molecules = ['C2S', 'HC7N_22_21', 'HC7N_21_20', 'HC5N']
+
+    	MolFile = '{0}/{0}_{2}_{1}.fits'.format(region,root,mol)
+	ConvFile = '{0}/{0}_{2}_{1}_conv.fits'.format(region,root,mol)
+	GaussOut = '{0}/{0}_{2}_{1}_gauss_cube.fits'.format(region,root,mol)
+	GaussNoiseOut = '{0}/{0}_{2}_{1}_gauss_cube_noise.fits'.format(region,root,mol)
+	ParamOut = '{0}/{0}_{2}_{1}_param_cube.fits'.format(region,root,mol)
+
+	# Load the spectral cube and convert to velocity units
+	cube = SpectralCube.read(MolFile)
+	cube_km = cube.with_spectral_unit(u.km / u.s, velocity_convention='radio')
+
+	# If desired, convolve map with larger beam 
+	# or load previously created convolved cube
+	if convolve!=False:
+		cube = SpectralCube.read(MolFile)
+		cube_km_1 = cube.with_spectral_unit(u.km / u.s, velocity_convention='radio')
+		beam = radio_beam.Beam(major=convolve*u.arcsec, minor=convolve*u.arcsec, pa=0*u.deg)
+		cube_km = cube_km_1.convolve_to(beam)
+		cube_km.write(ConvFile, format='fits', overwrite=True)
+	if use_old_conv!=False:
+		cube_km = SpectralCube.read(ConvFile)
+	
+	# Define the spectral axis in km/s
+	spectra_x_axis_kms = np.array(cube_km.spectral_axis) 
+
+	# Find the channel range corresponding to vmin and vmax
+	# -- This is a hold-over from when I originally set up the code to 
+	#    use a channel range rather than velocity range.
+	#    Can change later, but this should work for now. 
+	low_channel = np.where(spectra_x_axis_kms<=vmax)[0][0]+1 # Add ones to change index to channel
+	high_channel = np.where(spectra_x_axis_kms>=vmin)[0][-1]+1 # Again, hold-over from older setup
+	peak_channels = [low_channel, high_channel]
+
+	# Create cubes for storing the fitted Gaussian profiles
+	# and the Gaussians with noise added back into the spectrum
+	header = cube_km.header
+	cube_gauss = np.array(cube_km.unmasked_data[:,:,:])
+	cube_gauss_noise = np.array(cube_km.unmasked_data[:,:,:])
+	shape = np.shape(cube_gauss)
+
+	# Set up a cube for storing fitted parameters
+	param_cube = cube_gauss[0]
+	param_cube = param_cube.reshape((1,) + param_cube.shape)
+	param_cube = np.concatenate((param_cube, param_cube, param_cube, param_cube, param_cube, 	param_cube), axis=0)
+	param_header = cube_km.header
+
+	# Define the Gaussian profile
+	def p_eval(x, a, x0, sigma):
+    		return a*np.exp(-(x-x0)**2/(2*sigma**2))
+
+	# Create some arrays full of NANs
+	# To be used in output cubes if fits fail
+	nan_array=np.empty(shape[0]) # For gauss cubes
+	nan_array[:] = np.NAN
+	nan_array2=np.empty(param_cube.shape[0]) # For param cubes
+	nan_array2[:] = np.NAN
+
+	# Loop through each pixel and find those
+	# with SNR above snr_min
+	x = []
+	y = []
+	pixels = 0
+	for (i,j), value in np.ndenumerate(cube_gauss[0]):
+     		spectra=np.array(cube_km.unmasked_data[:,i,j])
+     		rms = np.std(np.append(spectra[0:(peak_channels[0]-1)], spectra[(peak_channels[1]+1):len(spectra)]))
+     		if (max(spectra[peak_channels[0]:peak_channels[1]]) / rms) > snr_min:
```

Comment:
Are you okay with this failing if there is a `nan` in the array?

Reply:
I've added a line to skip spectra that only contain nans and also changed to `np.nanstd` to catch spectra with some nans.